### DISCUSSION TITLE
Better Environment Variable workflow

### DISCUSSION THREAD

**Discussion by knolleary** [node-red/designs#77] [Design Proposals]
We use environment variables to make a flow more easily portable. However there are some gaps in the workflow that can make it quite awkward:

1. when creating a flow, not all properties can be set to an env var property as the node edit dialog doesn't provide a way to set one (eg boolean values).
2. when importing a flow that contains env vars, you do not get any help to know what env vars need to be set

I think there are things we could do to help here.

Some ideas for discussion:

### Creating flows with env vars

1. Provide a way to edit a node's raw JSON configuration directly - would allow *any* property to be set to an env var identifier. The downside is you lose some of the protection of the node's edit dialog to ensure the configuration is self-consistent and valid.
2. Alternatively, add a step in the Export Flow process that allows any node property to be replaced with an env var. Downside here is you would have to do that every time you exported an updated version.

### Importing a flow with env vars

1. When importing a flow, the import dialog could detect the presence of env vars and prompt the user to set some values - either on the containing Flow or as new Global env vars (see https://github.com/node-red/designs/discussions/76)

---

What else could we do to make env vars easier to work with?

**Comment by Steveorevo** [comment]
Not sure if I’m grasping the scenarios. is this for subflow import? 

With regards to #1, “downside is you lose some of the protection of the node's edit dialog…”. That seems paramount. If we wanted to design a particular node’s property to be env variable enabled; couldn’t we have chosen such behavior from the start? I see that we can design nodes specifically with that intent or to keep it fixed to a narrow set of data types and/or abilities. 

Perhaps that idea here is to introduce a new selector, an env variable *with* data type enforcement; as opposed to just an arbitrary env var? 

A boolean’s simplicity, or a checkbox is straightforward. If the intent here is to make it an env var enabled one too; perhaps a new, subtle difference in the control’s appearance could be created. Like a thin drop down arrow to the right hand side of the checkbox that reveals a hidden  field specifically for an env var entry input.

**Reply by knolleary** [reply]
> is this for subflow import?

For someone importing *any* flow into the editor, there is currently no support given to help them know if what they have just imported requires env vars to be set - they have to rely on being told, or poking around everything they just imported to see if there are any env var references.

For someone exporting a flow who wants to do a last-minute substitution of local configuration with env vars, they have to currently manually edit the JSON. I think we can provide help with that.

> If we wanted to design a particular node’s property to be env variable enabled; couldn’t we have chosen such behavior from the start

Yes, however there are 4000+ nodes in the community and the majority have given this no thought at all. So finding a solution that works in the general case is quite important.

**Comment by Steveorevo** [comment]
For "Importing a flow with environment variables"; that may be an ideal time to show a list of areas to be addressed; leveraging:

https://github.com/node-red/designs/discussions/79#discussion-4356962 "part of the url...fragment ends with /edit, [node's] edit dialog will be opened..."

**Comment by sammachin** [comment]
I'm not sure that opening up to modifying a nodes parameters directly in the JSON flow is a great idea, it feels like overstepping into the choices of the node-developer, we should be encouraging and making it easier for developers to use typed inputs (with env vars) in their nodes as best practice though.

**Reply by sammachin** [reply]
To elborate a little more on this my concern as a node developer would be the potential support overhead, if I have a field in my node that allows a user to select, `Red`, `Green` or `Blue` and the user now somehow manages to set this to `Badger` by modifying the JSON that could cause all kinds of strange errors

**Comment by HiroyasuNishiyama** [comment]
> What else could we do to make env vars easier to work with?

Considering a large-scale flow, it seems to be difficult for developers to change node properties directly.
Instead, I think developers should explicitly use environment variables to build reusable flows.

Following features may be useful for that purpose: 
- Specifying default value for environment variable like UNIX shell (e.g. ${VAR:=val}),
- Listing environment variables used within group or tab and adding, 
- Adding comment and a flag that distinguishes environment variable used as parameter.

**Comment by dceejay** [comment]
We already have several tools around Node-RED that may be relevant like NRlint or more probably the Flow Inspector (on the flows site)... While not directly helpful for exporting - if the concept of the inspector was brought into the core editor (maybe as - About flow ???) - it could trawl the flow and produce a report that included the list of used nodes - non-core nodes - environment variables - subflows - any other potentially useful parameters . (web endpoints ? urls ?) etc

**Comment by Steve-Mcl** [comment]
Hi all, I would like to resurrect this design and move it forward please.

I am particularly interested in ability to select a config node in the env var parameters of a subflow as this would greatly enhance the ability to encapsulate functionality (set flows) where nodes have config choices.

This becomes especially important in supporting Dashboard type nodes in a subflow.

Supporting this capability would open doors to OEMs who create templates (sometimes called widgetize or dynamos in other HMI/SCADA packages)

As I see it, there are 2 parts to this: 

1. Exposing the Config Select dropdown + edit button on the subflow instance
2. Providing a simple means of instructing the subflow template to use an ${env-var} instead of a fixed value

---

Part `1`: I have already demonstrated part `1` locally in my dev env (short video to follow)

Part `2`: requires further discussion and design but I see this similar to how @HiroyasuNishiyama does:

> * Listing environment variables used within group or tab and adding,
> * Adding comment and a flag that distinguishes environment variable used as parameter.

These 2 points align very closely with how I personally see this working.

1. Env Vars entered in the subflow are presented as a choice to the user when they chose a property to be supplied via env-var
2. The row is then flagged or highlighted as being overridden by an env-var

To make this clean, we might wish to consider introducing the feature via a new `data-xyz` element (in the form-row) that indicates to Node-RED this property is overridable.  Think of it as a contract with the core that the users has explicitly stated they have implemented the form-row in such a way it would behave correctly should the element be chosen to be overridden? 

Another consideration: Perhaps this is ONLY supported for form elements that are a `typedInput` or `config-select`.  This gives us the ability (at the core) to control how the property is rendered / operated when set to be an env var?


---

I intend on "having a play" with this over the coming weeks to see how it feels. I will post ideas / questions / demos here in the hope I can get feedback and resolution.

Cheers all.

**Comment by Steve-Mcl** [comment]
> Part `1`: I have already demonstrated part `1` locally in my dev env (short video to follow)

As promised, this is the working demo I have regarding support for Configs in a subflow.

https://github.com/node-red/designs/assets/44235289/32758eab-1eb5-4848-b396-73c6737b11d9

Comments (good and bad) most welcome.

**Reply by joepavitt** [reply]
SO good Steve ❤️

**Comment by Steve-Mcl** [comment]
And after some play around, a (very rough) visual demo of how we _might_ support simplified choice/selection of Env Vars

NOTES: 
* I have added the built-in `RED.contextMenu` to the first label of a `form-row` when the nodes `default`s are scanned and linked up to the UI elements
* The drop down presented is build from the internal knowledge that the node is INSIDE a subflow template.
   * the entries are built from the parent subflow templates' env-vars defined by the user
* Upon selecting an entry from the dropdown, I hide all internal elements of the `form-row` EXCEPT the label (this will NOT suit all node edit panels since many nodes are loose with the standard - this is where we _might_ need to introduce the feature via `data-` attributes or custom class names (still working that out))




https://github.com/node-red/designs/assets/44235289/334471c5-3859-4c54-9115-1b294e3cd4b2

**Reply by Steve-Mcl** [reply]
### Additional notes
* In this demo, the simple list is built from the raw names (not the friendly UI labels) - I would definitely want to pick these up


### Additional thoughts
Having had this swishing around my head, I have the additional thoughts:
*  I realise this UX (kinda) forces the user to first enter Env Vars.  We could have a menu item named `New` that inserts the new env var into the subflow config & enteres into the editing node.
* There is nothing to stop (or at minimum hint) what the expected value should be.  For example, the SIZE is not really a single number but a CSV number (e.g. "1,12") under the hood.
   * Permitting the user to add a hint in the Env Var setup might be a useful addition
* There is no validation that the Env Var exists. For example, a user may enter `${Group_ID}` instead of `${groupId}`
* The list does not present Globally defined Env Vars.
  * Personally, I am ok with this as it encourages more portable designs
* We could also display the icon in the menu (the icon chosen by the user in the Env Var setup) to aid faster recognition
* As Right Click is not entirely touch friendly, we may wish to support Double Click and/or have permanent icon buttons that permits the selection & override of a property?


### Missing bits:

Assuming we like this UX we need:
* a way of determining a property has been set to an env var so that when re-opening the node edit panel the correct UI is presented
* a way of remapping the new Env Var override element to the nodes `default` property

### Reality Check

In reality, applying this pattern carte blanche, I can foresee issues in simply hiding the original content and adding another element for the Env Var value. In particuar, nodes that have reactive elements based on user selection. I am now leaning towards a new, well defined UI pattern that is deliberately designed to support the "drop list" Env Var selection UX. Any node that currently employs a free-text entry can be utilised with the existing `${xxx}` syntax. 

I would hope that, where possible, the core nodes would be updated to take advantage of this "quick builder" Env Var selection.

### Stretch

in a group (or any node on a tab), would we want to offer the same level of support (Groups and Tags have Env Vars too!)

**Comment by joepavitt** [comment]
> I realise this UX (kinda) forces the user to first enter Env Vars. We could have a menu item named New that inserts the new env var into the subflow config & enteres into the editing node.

I don't think that's necessarily a bad thing, as a first pass of this, it certainly isn't a blocker imo

> The list does not present Globally defined Env Vars

Nor do I think it should, the scope here is about reusable subflows.

> We could also display the icon in the menu (the icon chosen by the user in the Env Var setup) to aid faster recognition

Great idea

> As Right Click is not entirely touch friendly, we may wish to support Double Click and/or have permanent icon buttons that permits the selection & override of a property?

I agree here, great for demonstration, but an icon of some kind would be preferred I think, but this may open concerns to a wider Node-RED sidebar aesthetics discussion

> in a group (or any node on a tab), would we want to offer the same level of support (Groups and Tags have Env Vars too!)

This being a subflow-only feature as a first pass is _plenty_

**Comment by Steve-Mcl** [comment]
Progress on the design - I will write up later what that entails but here is a teaser video of how I imagine we can make the override by Env Var much more accessible and RAD for OEMs

I would greatly appreciate feedback on the progress (would really like to know I am not wasting my time) - thank yoooo. :) 




https://github.com/node-red/designs/assets/44235289/e359aa03-46ba-4188-ac1c-e68c74c1720e

**Reply by Steve-Mcl** [reply]
Short version of how the fields of a node are instructed to be overridable.

More detail and reasoning and alternative considerations will be posted when I get time, but for now, this is the approach to supporting overridable elements in a nodes edit panel 

## Foreword

* Only nodes being edited in a subflow template are overridable
* Certain criterial about the structure of the for-row is enforced to minimise wrecking the UI 
* an additional defaults prop is required to store the overrides 
   * This was decided as the cleanest way to support importing a flow with overrides into an older node-red (since they will be ignored)

## How to enable override on a form row

### By data attribute

Applying the following attributes to the bound input (`node-input-*`) will instruct the init code to underline the adjacent label, append a link icon and become a context menu (triggered via left click)

| attr | about |
| --- | --- |
| `data-envvar-override="true"` | Inform the editor this input can be overriden |
| `data-envvar-override-name="Height"` | The friendly name to use in the auto completion <br> OPTIONAL - will use the associated default.prop name in its absence |
| `data-envvar-override-icon="fa fa-arrows-h"` | An icon to show in the auto completion <br> OPTIONAL - defaults to fa-tag |
| `data-envvar-override-i18n="widget-height"` | the i18n look up for the `name` <br> OPTIONAL - not implemented |


### By Definition in the defaults

```js
            defaults: {
                group: { type: 'ui-group', required: true },
                name: { get () { console.log('GET NAME'); return this._name }, set (v) { console.log('SETTING NAME'); this._name = v } },
                label: { value: 'text' },
                order: { value: 0 },
                height: {
                    value: 0,
                    ui: {
                        // autoGenerate: null // future feature proposal: allow UI to be generated from node def defaults e.g. { type: 'number', min: 0, max: 100, label: 'Height' }
                        envvarOverride: {
                            // enabled implicitly by the fact that this object exists
                            icon: 'font-awesome/fa-arrows-v',
                            label: 'Height'
                        }
                    }
                },
```


I hope that gives a flavour of the considerations made so far and spurns some conversation/questions/requests

All feedback welcomed.

**Reply by Steve-Mcl** [reply]
@joepavitt @knolleary @dceejay @HiroyasuNishiyama 

Firstly my apologies for the tag ^ and noise. 

If you guys could find 5 mins to mull over the progress and posts above (when you get a moment), I would greatly appreciate your thoughts?

Thank yooo.

**Reply by joepavitt** [reply]
Rather than "Override", possible an "Inherit" label instead?

**Reply by joepavitt** [reply]
<img width="204" alt="Screenshot 2023-12-01 at 10 31 38" src="https://github.com/node-red/designs/assets/99246719/f3011eb5-0ca0-4866-8d8c-db5b44714d13">

Bit confused by the two options here, does the bottom one effect the top?

**Reply by joepavitt** [reply]
Functionally, is there a difference between what you've got in `Label: ${text-label}` and the other that you configure with the dropdown option?

**Reply by Steve-Mcl** [reply]
> Functionally, is there a difference between what you've got in `Label: ${text-label}` and the other that you configure with the dropdown option?


Functionally, no. As per existing node-red capabilities, any entry with `${xxx}` will get the value of the env var.

**Reply by Steve-Mcl** [reply]
> <img alt="Screenshot 2023-12-01 at 10 31 38" width="204" src="https://private-user-images.githubusercontent.com/99246719/287237591-f3011eb5-0ca0-4866-8d8c-db5b44714d13.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE1MDYxNzgsIm5iZiI6MTcwMTUwNTg3OCwicGF0aCI6Ii85OTI0NjcxOS8yODcyMzc1OTEtZjMwMTFlYjUtMGNhMC00ODY2LThkOGMtZGI1YjQ0NzE0ZDEzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjAyVDA4MzExOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA1NWFkOTM1OTQ4MGUyNmQxZDI2NmFmNWJiZTIzNjcwODRkYTFmNTYxMzlmNDI0ZWNmNmRkNDZlZjU1ZDQ4YWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.WFRWNSik4ARFo3yeCtoxnNY1N82tNgmEcB4l_u8eKYQ">
> Bit confused by the two options here, does the bottom one effect the top?

Yes. But I have changed nothing in this area. Have a play with current node-red to see how these 2 interact.

**Reply by Steve-Mcl** [reply]
> Rather than "Override", possible an "Inherit" label instead?

Hmmm. Inherit sounds a little bit sexier but it has other connotations. I think override is more accurate and explicit. 

It could say "use env var" or something more friendly though.

**Reply by joepavitt** [reply]
> Inherit sounds a little bit sexier but it has other connotations

What other connotations? It's being inherited from the parent/wrapping Subflow, so it's accurate?

Override feels like the opposite, that the node would "override" any value provided by the parent

**Reply by Steve-Mcl** [reply]
Override was chosen from the perspective of the node. As in "the nodes property is being overridden". Since the users action is, at that instant in time, operating on an existing node property.

I'm my mind saying "the node property is inherited" made me immediately think of OOO inheritance (like a class inherits a property from a parent class, not the value, the property itself).

But you make fair points & I am almost convinced, but i just don't feel comfortable with "inherit".

Happy to be ~~inherited~~ overridden if anyone else ever joins the thread 😉

**Reply by svdasein** [reply]
@Steve-Mcl hi - I'm just curious as to how this thing is progressing?  It'd be an awesome addition to node-red! Any idea when it might make its way into a release?
